Antivirus (#5846)
owncloud/ocis: services/antivirus/README.md

> # Antivirus Service
> 
> The `antivirus` service is responsible for scanning files for viruses
> 
> ## Configuration
> 
> ### Antivirus Scanner Type
> 
> The antivirus service currently supports `icap` and `clamav` as antivirus scanners. Use `ANTIVIRUS_SCANNER_TYPE` to configure this. 
> Note that configuration depends heavily on chosen antivirus scanner. See Enviroment Variable descriptions for details.
> 
> ### Maximum Scan size
> 
> Since several factors might make need necessary to limit the maximum filesize the `antivirus` service has an option to set a max scan size.
> Use `ANTIVIRUS_MAX_SCAN_SIZE` to scan only that amount of bytes of a file. Obviously it is recommended to set this as high as possible, but several factors (scanner type and version, bandwith and performance issues, ...) might force to set this to a certain filesize.
> 
> ### Infected File Handling
> 
> The `antivirus` service allows three different ways of handling infected files. Those can be set via the `ANTIVIRUS_INFECTED_FILE_HANDLING` envvar:
>   -   `delete` (default): Infected files will be deleted immediately. Further postprocessing is cancelled.
>   -   `abort`: Infected files will be kept. Further postprocessing is cancelled. Files can be manually retrieved and inspected by an admin. (Advanced option)
>   -   `continue`: Infected files will be marked as infected but postprocessing continues normally. Note: Infected Files are not prevented from download. Risk of spreading viruses. (Obviously not recommended)
> 
> ## Operation Modes
> 
> The `antivirus` service can scan files during postprocessing. `on demand` scanning will be added in the future.
> 
> ### Postprocessing
> 
> Note: Needs to be configured via the [postprocessing service](https://github.com/owncloud/ocis/tree/master/services/postprocessing) 
> 
> The `antivirus` service will scan files during postprocessing. It listens for a postprocessing step called `"virusscan"`
> 
> ### On Demand
> 
> On demand scanning is currently not supported

Reviewer: ```suggestion
# Antivirus Service

The `antivirus` service is responsible for scanning files for viruses.

## Configuration

### Antivirus Scanner Type

The antivirus service currently supports `icap` and `clamav` as antivirus scanners. The `ANTIVIRUS_SCANNER_TYPE` environment variable is used to select the scanner. The detailed configuration for each scanner heavily depends on the scanner type selected. See the environment variables for more details.

### Maximum Scan size

Several factors might make it necessary to limit the maximum filesize the antivirus service will scan. Use the `ANTIVIRUS_MAX_SCAN_SIZE` environment variable to scan only the amount of bytes of a file defined. Obviously it is recommended to set this as high as possible, but several factors like scanner type and version, bandwith, performance issues, etc.) might force to limit this to a certain filesize.

### Infected File Handling

The antivirus service allows three different ways of handling infected files. Those can be set via the `ANTIVIRUS_INFECTED_FILE_HANDLING` environment variable:

  -   `delete` (default): Infected files will be deleted immediately. Further postprocessing is cancelled.
  -   `abort`:  (Advanced option): Infected files will be kept. Further postprocessing is cancelled. Files can be manually retrieved and inspected by an admin. To identify the file for further investigation, the antivirus services logs the abort/infected state including the file ID. This file is located in the `storage/users/uploads` folder of the ocis data directory and persists until it is manually deleted by the admin via the [Manage Unfinished Uploads](https://doc.owncloud.com/ocis/next/deployment/services/s-list/storage-users.html#manage-unfinished-uploads) command.
  -   `continue`:  (Obviously not recommended): Infected files will be marked as infected but postprocessing continues normally. Note: Infected Files are not prevented from download. Risk of spreading viruses.

## Operation Modes

The antivirus service can scan files during postprocessing. `on demand` scanning might be added in the future.

### Postprocessing

The antivirus service will scan files during postprocessing. It listens for a postprocessing step called `"virusscan"`. This step can be added in the environment variable `POSTPROCESSING_STEPS`.

Read the documentation of the [postprocessing service](https://github.com/owncloud/ocis/tree/master/services/postprocessing) for more details.
```

# Antivirus Service

The `antivirus` service is responsible for scanning files for viruses.

## Configuration

### Antivirus Scanner Type

The antivirus service currently supports [icap](https://tools.ietf.org/html/rfc3507) and [clamav](http://www.clamav.net/index.html) as antivirus scanners. The `ANTIVIRUS_SCANNER_TYPE` environment variable is used to select the scanner. The detailed configuration for each scanner heavily depends on the scanner type selected. See the environment variables for more details.

  -   For `icap`, only scanners using the `X-Infection-Found` header are currently supported.
  -   For `clamav` only local sockets can currently be configured.

### Maximum Scan size

> Several factors might make it necessary to limit the maximum filesize the antivirus service will use for scanning. Use the `ANTIVIRUS_MAX_SCAN_SIZE` environment variable to scan only the amount of bytes of a file defined. Obviously it is recommended to set this as high as possible, but several factors like scanner type and version, bandwith, performance issues, etc. might force to limit this to a certain filesize.

Reviewer: ```suggestion
Several factors can make it necessary to limit the maximum filesize the antivirus service will use for scanning. Use the `ANTIVIRUS_MAX_SCAN_SIZE` environment variable to scan only a given amount of bytes. Obviously it is recommended to scan the whole file, but several factors like scanner type and version, bandwith, performance issues, etc. might make a limit necessary.
```
